File: lib/analysis_engine/utils/formatting.js
export function money(value, symbol = "$") {
  return `${symbol}${value.toLocaleString("en-US", { maximumFractionDigits: 0 })}`;
}

export function compact_money(value, symbol = "$") {
  const abs_value = Math.abs(value);
  if (abs_value >= 1_000_000_000) return `${symbol}${(value / 1_000_000_000).toFixed(1)}B`;
  if (abs_value >= 1_000_000) return `${symbol}${(value / 1_000_000).toFixed(1)}M`;
  return money(value, symbol);
}

export function pct(value) {
  if (value == null) return "n/a";
  return `${(value * 100).toLocaleString("en-US", { maximumFractionDigits: 1 })}%`;
}

File: analysis_engine/utils/formatting.py
"""Small display formatting helpers for story prototypes."""

from __future__ import annotations

from typing import Optional


def money(value: float, symbol: str = "$") -> str:
    """Format a number as whole-unit currency."""
    return f"{symbol}{value:,.0f}"


def compact_money(value: float, symbol: str = "$") -> str:
    """Format large currency values with B/M suffixes."""
    if abs(value) >= 1_000_000_000:
        return f"{symbol}{value / 1_000_000_000:,.1f}B"
    if abs(value) >= 1_000_000:
        return f"{symbol}{value / 1_000_000:,.1f}M"
    return money(value, symbol=symbol)


def pct(value: Optional[float]) -> str:
    """Format a decimal return or ratio as a percentage."""
    return "n/a" if value is None else f"{value * 100:,.1f}%"
